lntemp/node/watcher.go

```diff
@@ -0,0 +1,679 @@
+		if err != nil {
+			return err
+		}
+
+		return fmt.Errorf("channel:%s not heard before timeout: "+
+			"node has heard: %s", op, updates)
+	}
+}
+
+// WaitForChannelClose will block until a channel with the target outpoint is
+// seen as closed within the node's network topology. A channel is considered
+// closed once a transaction spending the funding outpoint is seen within a
+// confirmed block.
+func (nw *nodeWatcher) WaitForChannelClose(
+	chanPoint *lnrpc.ChannelPoint) (*lnrpc.ClosedChannelUpdate, error) {
+
+	op := nw.rpc.MakeOutpoint(chanPoint)
+	eventChan := make(chan struct{})
+	nw.chanWatchRequests <- &chanWatchRequest{
+		chanPoint:     op,
+		eventChan:     eventChan,
+		chanWatchType: watchCloseChannel,
+	}
+
+	timer := time.After(DefaultTimeout)
+	select {
+	case <-eventChan:
+		closedChan, ok := nw.state.closedChans.Load(op)
+		if !ok {
+			return nil, fmt.Errorf("channel:%s expected to find "+
+				"a closed channel in node's state:%s", op,
+				nw.state)
+		}
+		return closedChan.(*lnrpc.ClosedChannelUpdate), nil
+
+	case <-timer:
+		return nil, fmt.Errorf("channel:%s not closed before timeout: "+
+			"%s", op, nw.state)
+	}
+}
+
+// syncMapToJSON is a helper function that creates json bytes from the sync.Map
+// used in the node. Expect the sync.Map to have map[string]interface.
+func syncMapToJSON(state *sync.Map) ([]byte, error) {
+	m := map[string]interface{}{}
+	state.Range(func(k, v interface{}) bool {
+		op := k.(wire.OutPoint)
+		m[op.String()] = v
+		return true
+	})
+	policies, err := json.MarshalIndent(m, "", "\t")
+	if err != nil {
+		return nil, fmt.Errorf("encode polices err: %v", err)
+	}
+
+	return policies, nil
+}
+
+// topologyWatcher is a goroutine which is able to dispatch notifications once
+// it has been observed that a target channel has been closed or opened within
+// the network. In order to dispatch these notifications, the
+// GraphTopologySubscription client exposed as part of the gRPC interface is
+// used.
+//
+// NOTE: must be run as a goroutine.
+func (nw *nodeWatcher) topologyWatcher(ctxb context.Context,
+	started chan error) {
+
+	graphUpdates := make(chan *lnrpc.GraphTopologyUpdate)
+
+	client, err := nw.newTopologyClient(ctxb)
+	started <- err
+
+	// Exit if there's an error.
+	if err != nil {
+		return
+	}
+
+	// Start a goroutine to receive graph updates.
+	nw.wg.Add(1)
+	go func() {
+		defer nw.wg.Done()
+
+		// With the client being created, we now start receiving the
+		// updates.
+		err = nw.receiveTopologyClientStream(ctxb, client, graphUpdates)
+		if err != nil {
+			started <- fmt.Errorf("receiveTopologyClientStream "+
+				"got err: %v", err)
+		}
+	}()
+
+	for {
+		select {
+		// A new graph update has just been received, so we'll examine
+		// the current set of registered clients to see if we can
+		// dispatch any requests.
+		case graphUpdate := <-graphUpdates:
+			nw.handleChannelEdgeUpdates(graphUpdate.ChannelUpdates)
```

Comment:
Looks like this and the below can be wrapped to 80 columns?

Reply:
I think they are?